go-controller/pkg/ovn/controller/services/lb_config.go

```diff
@@ -92,9 +93,18 @@ func makeNodeRouterTargetIPs(node *nodeInfo, c *lbConfig, hostMasqueradeIPV4, ho
 	}
 
 	// Any targets local to the node need to have a special
-	// harpin IP added, but only for the router LB
-	targetIPsV4, v4Updated := util.UpdateIPsSlice(targetIPsV4, node.l3gatewayAddressesStr(), []string{hostMasqueradeIPV4})
-	targetIPsV6, v6Updated := util.UpdateIPsSlice(targetIPsV6, node.l3gatewayAddressesStr(), []string{hostMasqueradeIPV6})
+	// hairpin IP added, but only for the router LB
+	hostMasqueradeIPsV4 := node.l3gatewayAddressesStr()
+	hostMasqueradeIPsV6 := node.l3gatewayAddressesStr()
+
+	if config.OvnKubeNode.Mode == types.NodeModeDPU {
+		// hairpin traffic of the hostaddress in case of the DPU
+		hostMasqueradeIPsV4 = node.hostAddressesStr()
+		hostMasqueradeIPsV6 = node.hostAddressesStr()
+	}
+
+	targetIPsV4, v4Updated = util.UpdateIPsSlice(targetIPsV4, hostMasqueradeIPsV4, []string{hostMasqueradeIPV4})
+	targetIPsV6, v6Updated = util.UpdateIPsSlice(targetIPsV6, hostMasqueradeIPsV6, []string{hostMasqueradeIPV6})
```

Comment:
so really what you are doing here is trying to ignore using the l3 gateway address and include other host ips to find and replace addresses with masquerade IPs for the GR loadbalancer VIP. In shared gateway mode we assume that the gwIntf ==kube api IP and that assumption is why you need to change this code. I think your change is valid even in a *non* DPU case. I would:

1. remove this function: node.l3gatewayAddressesStr()
2. always use node.hostAddressesStr() for UpdateIPSlice

Then any IP used on the host will be substituted for the masquerade IP, which what needs to happen to make host->service work for any interface that is used as KAPI.